onnxruntime/core/providers/cpu/ml/tree_ensemble_classifier.cc

```diff
@@ -358,20 +358,23 @@ common::Status TreeEnsembleClassifier<T>::Compute(OpKernelContext* context) cons
   int64_t N = x_dims.size() == 1 ? 1 : x_dims[0];
   Tensor* Y = context->Output(0, TensorShape({N}));
   auto* Z = context->Output(1, TensorShape({N, class_count_}));
-
-  int64_t zindex = 0;
   const T* x_data = X.template Data<T>();
 
-  // for each class
-  std::vector<float> scores;
-  scores.reserve(class_count_);
+  common::Status status;
+#ifdef USE_OPENMP
+#pragma omp parallel for
+#endif
   for (int64_t i = 0; i < N; ++i) {
-    scores.clear();
+    int64_t zindex = i * class_count_;
+    std::vector<float> scores;
     int64_t current_weight_0 = i * stride;
     std::map<int64_t, float> classes;
     // walk each tree from its root
     for (size_t j = 0, end = roots_.size(); j < end; ++j) {
-      ORT_RETURN_IF_ERROR(ProcessTreeNode(classes, roots_[j], x_data, current_weight_0));
+      auto process_status = ProcessTreeNode(classes, roots_[j], x_data, current_weight_0);
```

Comment:
This function makes executes classes.insert(p1); (classes is not protected against multithreading). Not sure OMP is able to detect it.

Reply:
Good point - not targeting that as of yet. 
Without collapsing, openmp will not parallelize nested loop.